### PR TITLE
.NET Client

### PR DESCRIPTION
**Pull request by mikaelm12**
Re-Opening this for further review.
Tests have been added since the .NET client was last looked at. One thing to note is that we need a solution for the send before poll issue. 
@anurse @moozzyk @BrennanConroy @davidfowl

### CODE REVIEW

File: samples/SocketsSample/LineInvocationAdapter.cs [hunk 1]
```diff
@@ -22,31 +23,61 @@ public class LineInvocationAdapter : IInvocationAdapter
...
+                    Method = method,
+                    Arguments = values.Skip(2).Zip(binder.GetParameterTypes(method), (v, t) => Convert.ChangeType(v, t)).ToArray()
+                };
+            }
+        }
+
+        public Task WriteMessageAsync(InvocationMessage message, Stream stream, CancellationToken cancellationToken)
+        {
+            var invocationDescriptor = message as InvocationDescriptor;
+            if(invocationDescriptor != null)
```
Comment (BrennanConroy) [inline samples/SocketsSample/LineInvocationAdapter.cs]: nit: space

File: src/Microsoft.AspNetCore.SignalR.Common/JsonNetInvocationAdapter.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    public class JsonNetInvocationAdapter : IInvocationAdapter
+    {
+        private JsonSerializer _serializer = new JsonSerializer();
+
+        public JsonNetInvocationAdapter()
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Common/JsonNetInvocationAdapter.cs]: Is the empty constructor necessary?

File: src/Microsoft.AspNetCore.SignalR.Common/JsonNetInvocationAdapter.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                Id = json.Value<string>("Id"),
+                Method = json.Value<string>("Method"),
+            };
+
+            var paramTypes = binder.GetParameterTypes(invocation.Method);
+            invocation.Arguments = new object[paramTypes.Length];
+
+            var args = json.Value<JArray>("Arguments");
+            for (int i = 0; i < paramTypes.Length; i++)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Common/JsonNetInvocationAdapter.cs]: nit: var

File: samples/ClientSample/HubSample.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        public static async Task MainAsync(string[] args)
+        {
+            var baseUrl = "http://localhost:5000/hubs";
+            if (args.Length > 0)
+            {
+                baseUrl = args[0];
+            }
+
+            var loggerFactory = new LoggerFactory();
```
Comment (anurse) [inline samples/ClientSample/HubSample.cs]: No argument here. Wanted to start with everything I thought we'd need though so it's easier to clean-up.

File: src/Microsoft.AspNetCore.Sockets.Client/BufferContent.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿using System.IO;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client/BufferContent.cs]: Copyright

File: src/Microsoft.AspNetCore.Sockets.Client/Connection.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
+﻿using System;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client/Connection.cs]: Copyright

File: src/Microsoft.AspNetCore.Sockets.Client/project.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      "CS1591"
+    ],
+    "xmlDoc": true
+  },
+
+  "dependencies": {
+    "NETStandard.Library": "1.6.1-*",
+    "System.IO.Pipelines": "0.1.0-*",
+    "Microsoft.Extensions.Logging.Abstractions": "1.2.0-*",
+    "Microsoft.Extensions.TaskCache.Sources": "1.2.0-*"
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client/project.json]: build reference

File: test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/project.json [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    "dotnet-test-xunit": "2.2.0-*",
+    "Microsoft.AspNetCore.Http": "1.1.0",
+    "Microsoft.AspNetCore.Sockets": "0.1.0-*",
+    "Microsoft.AspNetCore.Hosting": "1.1.0-*",
+    "Microsoft.AspNetCore.Server.Kestrel": "1.1.0-*",
+    "Microsoft.AspNetCore.SignalR": {
+      "target": "project"
+    },
+    "Microsoft.AspNetCore.SignalR.Client": {
+      "target": "project"
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/project.json]: Don't do project reference in test projects

Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/project.json]: Yeah, even though it sounds like the right thing to do, Universe removes the source projects in order to force the tests to run against packages, so this constraint will cause a problem.

Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/project.json]: Lesson learned

**Comment by mikaelm12** [comment]
⬆️ 📅 @BrennanConroy @moozzyk @anurse @davidfowl

**Comment by moozzyk** [comment]
:ship::it:

File: test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            var loggerFactory = new LoggerFactory();
+            const string originalMessage = "SignalR";
+
+            using (var httpClient = _testServer.CreateClient())
+            using (var pipelineFactory = new PipelineFactory())
+            {
+                var transport = new LongPollingTransport(httpClient, loggerFactory);
+                using (var connection = await HubConnection.ConnectAsync(new Uri("http://test/hubs"), new JsonNetInvocationAdapter(), transport, httpClient, pipelineFactory, loggerFactory))
+                {
+                    //TODO: Get rid of this. This is to prevent "No channel" failures due to sends occuring before the first poll.
```
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs]: Is there an issue tracking this?

File: SignalR.sln [hunk 10]
```diff
@@ -45,6 +43,30 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNetCore.Signal
...
+EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNetCore.SignalR.Client", "src\Microsoft.AspNetCore.SignalR.Client\Microsoft.AspNetCore.SignalR.Client.xproj", "{354335AB-CEE9-4434-A641-78058F6EFE56}"
+EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNetCore.SignalR.Client.FunctionalTests", "test\Microsoft.AspNetCore.SignalR.Client.FunctionalTests\Microsoft.AspNetCore.SignalR.Client.FunctionalTests.xproj", "{455B68D2-C5B6-4BF4-A685-964B07AFAAF8}"
+EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNetCore.Server.IntegrationTesting", "..\Hosting\src\Microsoft.AspNetCore.Server.IntegrationTesting\Microsoft.AspNetCore.Server.IntegrationTesting.xproj", "{3DA89347-6731-4366-80C4-548F24E8607B}"
+EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNetCore.Hosting.Abstractions", "..\Hosting\src\Microsoft.AspNetCore.Hosting.Abstractions\Microsoft.AspNetCore.Hosting.Abstractions.xproj", "{BB780FBB-7842-4759-8DE7-96FA2E5571C1}"
```
Comment (JunTaoLuo) [inline SignalR.sln]: Remove these packages pulled in via global.json

File: test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs (hunk 9 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs]: Yes. https://github.com/aspnet/SignalR/issues/57

**Comment by JunTaoLuo** [comment]
Many files are missing copyright...

**Pull request merged**